servlet/src/main/java/io/undertow/servlet/handlers/security/SSLInformationAssociationHandler.java

```diff
@@ -96,23 +94,31 @@ public static int getKeyLength(String cipherSuite) {
      */
     private X509Certificate[] getCerts(SSLSessionInfo session) {
         try {
-            javax.security.cert.X509Certificate[] javaxCerts = session.getPeerCertificateChain();
-            if (javaxCerts == null || javaxCerts.length == 0) {
+            Certificate[] javaCerts = session.getPeerCertificates();
+            if (javaCerts == null) {
                 return null;
             }
-            X509Certificate[] javaCerts = new X509Certificate[javaxCerts.length];
-            java.security.cert.CertificateFactory cf = java.security.cert.CertificateFactory.getInstance("X.509");
-            for (int i = 0; i < javaxCerts.length; i++) {
-                byte[] bytes = javaxCerts[i].getEncoded();
-                ByteArrayInputStream stream = new ByteArrayInputStream(bytes);
-                javaCerts[i] = (X509Certificate) cf.generateCertificate(stream);
+            int x509Certs = 0;
+            for (Certificate javaCert : javaCerts) {
+                if (javaCert instanceof X509Certificate) {
+                    ++x509Certs;
+                }
             }
-
-            return javaCerts;
+            if (x509Certs == 0) {
+                return null;
+            }
+            if (x509Certs == javaCerts.length) {
+                return (X509Certificate[]) javaCerts;
+            }
+            X509Certificate[] results = new X509Certificate[x509Certs];
+            for (int i = 0; i < x509Certs; i++) {
+                results[i] = (X509Certificate) javaCerts[i];
```

Comment:
You cannot be sure certificate on index i is instance of X509Certificate.
The loop should be using instanceof check.

Reply:
oh geez, I had a dumb... fix incoming